### PR TITLE
Preserving symlink or not can be configured from Command.

### PR DESCRIPTION
**Pull request by yoshisatoyanagisawa**
For some toolchains, a user need to run a wrapper to run a compiler.  Such a wrapper sometimes use readlink inside and resolving symlink breaks such a script.  To make it work, we need to preserve symlinks.  Since the behavior is only needed for certain toolchains, let me make it configurable via Command.InputSpec message.

### CODE REVIEW

File: go/api/command/command.proto [hunk 1]
```diff
@@ -87,6 +87,17 @@ message VirtualInput {
   bool is_empty_directory = 4;
 }
 
+message SymlinkBehaviorType {
+  enum Value {
```
Comment (atetubou) [inline go/api/command/command.proto:91]: Why do you need this nest?

Also follow
https://developers.google.com/protocol-buffers/docs/style#enums
?

Comment (yoshisatoyanagisawa) [reply go/api/command/command.proto:91]: I followed how other enums are configured.  To be honest, I do not understand why this is needed.  I just tried to guess the coding style here from CommandResultStatus and InputType.

**Comment by yoshisatoyanagisawa** [comment]
I focused so much on the main logic X(
I will add tests right away.

File: go/pkg/command/command.go [hunk 2]
```diff
@@ -45,6 +45,33 @@ func (s InputType) String() string {
 	return fmt.Sprintf("InvalidInputType(%d)", s)
 }
 
+// SymlinkBehavior represents how symlinks are handled.
```
Comment (atetubou) [inline go/pkg/command/command.go:48]: SymlinkBehaviorType?

File: go/pkg/command/command.go [hunk 3]
```diff
@@ -45,6 +45,33 @@ func (s InputType) String() string {
...
+	UnspecifiedSymlinkBehavior SymlinkBehaviorType = iota
+
+	// ResolveSymlink means symlinks are resolved.
+	ResolveSymlink
+
+	// PreserveSymlink means symlinks are kept as-is.
+	PreserveSymlink
+
+	// SymlinkBehaviorSentinel represents the end of SymlinkBehavior.
+	SymlinkBehaviorSentinel
```
Comment (atetubou) [inline go/pkg/command/command.go:63]: no needs to be public?

File: go/pkg/command/command.go [hunk 4]
```diff
@@ -45,6 +45,33 @@ func (s InputType) String() string {
...
+)
+
+var symlinkBehaviorType = [...]string{"UnspecifiedSymlinkBehavior", "ResolveSymlink", "PreserveSymlink"}
+
+func (s SymlinkBehaviorType) String() string {
+	if UnspecifiedSymlinkBehavior <= s && s < SymlinkBehaviorSentinel {
```
Comment (atetubou) [inline go/pkg/command/command.go:69]:
```suggestion
	if UnspecifiedSymlinkBehavior <= s && s < len(symlinkBehaviorType) {
```

File: go/pkg/command/command.go [hunk 5]
```diff
@@ -45,6 +45,33 @@ func (s InputType) String() string {
 	return fmt.Sprintf("InvalidInputType(%d)", s)
 }
 
+// SymlinkBehavior represents how symlinks are handled.
+type SymlinkBehaviorType int
+
+const (
+	// UnspecifiedSymlinkBehavior means following clients.TreeSymlinkOpts
+	// or DefaultTreeSymlinkOpts if clients.TreeSymlinkOpts is null.
+	UnspecifiedSymlinkBehavior SymlinkBehaviorType = iota
```
Comment (atetubou) [inline go/pkg/command/command.go:54]: no needs to be public?

File: go/pkg/command/command.go (hunk 3 above)
Comment (yoshisatoyanagisawa) [reply go/pkg/command/command.go:63]: good catch.

File: go/pkg/command/command.go (hunk 4 above)
Comment (yoshisatoyanagisawa) [reply go/pkg/command/command.go:69]: yeah, that sounds better.

File: go/pkg/command/command.go (hunk 5 above)
Comment (yoshisatoyanagisawa) [reply go/pkg/command/command.go:54]: I am not so comfortable to reveal a value nobody outside of the package can compare.  Let me defer to the repository owner for the decision.

File: go/pkg/command/command.go [hunk 6]
```diff
@@ -45,6 +45,30 @@ func (s InputType) String() string {
...
+	ResolveSymlink
+
+	// PreserveSymlink means symlinks are kept as-is.
+	PreserveSymlink
+)
+
+var symlinkBehaviorType = [...]string{"UnspecifiedSymlinkBehavior", "ResolveSymlink", "PreserveSymlink"}
+
+func (s SymlinkBehaviorType) String() string {
+	if UnspecifiedSymlinkBehavior <= s && s < SymlinkBehaviorType(len(symlinkBehaviorType)) {
```
Comment (atetubou) [inline go/pkg/command/command.go:66]: better to cast `s` to `int`?

File: go/pkg/command/command.go (hunk 5 above)
Comment (ramymedhat) [reply go/pkg/command/command.go:54]: I think this being public is fine and consistent with the rest of the file.

File: go/api/command/command.proto [hunk 7]
```diff
@@ -87,6 +87,17 @@ message VirtualInput {
...
+    // Use Client.TreeSymlinkOpts or default if it is not set.
+    UNSPECIFIED = 0;
+    // Resolves symlinks.
+    RESOLVE = 1;
```
Comment (rubensf) [inline go/api/command/command.proto:95]: I think we should just leave the default as "resolve" (eg resolve = 0). On inputs, unspecified actually has a meaning.

Also I think "follow" might be a more descriptive name.

File: go/api/command/command.proto [hunk 8]
```diff
@@ -104,6 +115,9 @@ message InputSpec {
 
   // Environment variables the command relies on.
   map<string,string> environment_variables = 4;
+
+  // symlinks are handled in this way.
```
[the remark this answers is not in the document]
Comment (rubensf) [reply go/api/command/command.proto:119]: And start sentence with capital letters.

File: go/pkg/command/command.go (hunk 2 above)
Comment (yoshisatoyanagisawa) [reply go/pkg/command/command.go]: good catch.

File: go/pkg/command/command.go (hunk 6 above)
Comment (yoshisatoyanagisawa) [reply go/pkg/command/command.go]: Hmm, instead of trying to write something different, let me follow how InputType is written.

**Comment by yoshisatoyanagisawa** [comment]
PTAL

File: go/api/command/command.proto (hunk 7 above)
Comment (rubensf) [reply go/api/command/command.proto:95]: Ah, my bad, I think you're right on both accounts!

**Review by rubensf — APPROVED** [review]

**Comment by yoshisatoyanagisawa** [comment]
I do not have  a write access to this repository.
Will somebody with the permission merge this?

Thank you.

**Pull request merged**